File: python/packages/realtime-core/realtime_core/frames.py
"""WebSocket frame encoders + inbound parser for the realtime service.

Mirrors the realtime server's protocol message definitions. Server-published events arrive as
{"type": "message", "channel": str, "data": {"event": str, "payload": dict}}
— the frame `type` is "message" (ServerMessageType.MESSAGE), NOT "event".
The inner data.event is the app-level event name (e.g. "run.succeeded")."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


def subscribe_frame(channel: str, *, scope: str | None = None, msg_id: str | None = None) -> dict[str, Any]:
    frame: dict[str, Any] = {"type": "subscribe", "channel": channel}
    if scope is not None:
        frame["scope"] = scope
    if msg_id is not None:
        frame["id"] = msg_id
    return frame


def unsubscribe_frame(channel: str) -> dict[str, Any]:
    return {"type": "unsubscribe", "channel": channel}


def publish_frame(channel: str, data: dict[str, Any], *, scope: str | None = None) -> dict[str, Any]:
    frame: dict[str, Any] = {"type": "publish", "channel": channel, "data": data}
    if scope is not None:
        frame["scope"] = scope
    return frame


def ping_frame() -> dict[str, Any]:
    return {"type": "ping"}


@dataclass(frozen=True)
class EventFrame:
    channel: str
    event: str
    payload: dict[str, Any]
    sender_id: str = ""


def parse_inbound(msg: dict[str, Any]) -> EventFrame | None:
    if msg.get("type") != "message":
        return None
    channel = msg.get("channel")
    if not channel:
        return None
    data = msg.get("data") or {}
    return EventFrame(
        channel=channel,
        event=data.get("event", ""),
        payload=data.get("payload") or {},
        sender_id=msg.get("sender_id", ""),
    )


# Inbound frame `type` values that carry channel data the SDK surfaces:
# application messages plus the full presence lifecycle.
_INBOUND_FRAME_TYPES = frozenset(
    {"message", "presence:state", "presence:join", "presence:leave", "presence:update"}
)


@dataclass(frozen=True)
class InboundEvent:
    """A raw inbound frame the SDK surfaces verbatim.

    `kind` is the frame `type` ("message" or any "presence:*"); `data` is the
    RAW inner data dict, NOT assumed to be {event, payload}."""

    kind: str
    channel: str
    sender_id: str
    data: dict[str, Any]


def parse_frame(msg: dict[str, Any]) -> InboundEvent | None:
    """Parse any surfaced inbound frame (message + presence) into an InboundEvent.

    Returns None for control/transport frames (pong, subscription_succeeded,
    errors, etc.)."""
    kind = msg.get("type")
    if kind not in _INBOUND_FRAME_TYPES:
        return None
    return InboundEvent(
        kind=kind,
        channel=msg.get("channel", ""),
        sender_id=msg.get("sender_id", ""),
        data=msg.get("data") or {},
    )

File: dotnet/src/BrutalSystems.Realtime.Core/Frames.cs
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace BrutalSystems.Realtime.Core;

public sealed record EventFrame(string Channel, string Event, JsonObject Payload, string SenderId = "");

public sealed record InboundEvent(string Kind, string Channel, string SenderId, JsonObject Data);

public static class Frames
{
    private static readonly HashSet<string> InboundTypes = new()
    {
        "message", "presence:state", "presence:join", "presence:leave", "presence:update",
    };

    public static JsonObject Subscribe(string channel, string? scope = null, string? id = null)
    {
        var f = new JsonObject { ["type"] = "subscribe", ["channel"] = channel };
        if (scope is not null) f["scope"] = scope;
        if (id is not null) f["id"] = id;
        return f;
    }

    public static JsonObject Unsubscribe(string channel) =>
        new() { ["type"] = "unsubscribe", ["channel"] = channel };

    public static JsonObject Publish(string channel, object data, string? scope = null, string? id = null)
    {
        var f = new JsonObject
        {
            ["type"] = "publish",
            ["channel"] = channel,
            // Round-trip `data` through the Web serializer into a JsonNode so
            // anonymous objects land as a JSON object with their own key casing.
            ["data"] = JsonSerializer.SerializeToNode(data, WebOpts),
        };
        if (scope is not null) f["scope"] = scope;
        if (id is not null) f["id"] = id;
        return f;
    }

    public static JsonObject Ping() => new() { ["type"] = "ping" };

    public static EventFrame? ParseInbound(JsonObject msg)
    {
        if (msg["type"]?.GetValue<string>() != "message") return null;
        var channel = msg["channel"]?.GetValue<string>();
        if (string.IsNullOrEmpty(channel)) return null;
        var data = msg["data"] as JsonObject ?? new JsonObject();
        return new EventFrame(
            Channel: channel,
            Event: data["event"]?.GetValue<string>() ?? "",
            Payload: data["payload"] as JsonObject ?? new JsonObject(),
            SenderId: msg["sender_id"]?.GetValue<string>() ?? "");
    }

    public static InboundEvent? ParseFrame(JsonObject msg)
    {
        var kind = msg["type"]?.GetValue<string>();
        if (kind is null || !InboundTypes.Contains(kind)) return null;
        return new InboundEvent(
            Kind: kind,
            Channel: msg["channel"]?.GetValue<string>() ?? "",
            SenderId: msg["sender_id"]?.GetValue<string>() ?? "",
            Data: msg["data"] as JsonObject ?? new JsonObject());
    }

    private static readonly JsonSerializerOptions WebOpts = new(JsonSerializerDefaults.Web);
}
